Distance sensor capabilities

xabierolaz opened the pull request:
### Requirements for Pull Request
* Read [CONTRIBUTING.md](https://github.com/wil3/gymfc/blob/master/CONTRIBUTING.md).
* Open an issue. Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of the pull request.
* Fill out the template below, any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.

### Description of the Change


#### 1-Added message definition for the sensor by updating the state protobuff message


### Verification Process
<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (commands you ran, etc.), and describe the results you observed.

-->
Has not been verified yet, as it's just the capability of adding the distance sensor into gymfc

Tested by running `test_axis.py`

No errors in console. However, sensor does not seem to appear in Gazebo yet inside the nf1 model

xabierolaz committed:
First commit

1-Updated entry in sdf model

2-Added message definition for the sensor by updating the state protobuff message

wil3 commented on examples/gymfc_nf/twins/nf1/model.sdf:
```diff
@@ -739,6 +739,38 @@
+          <stddev>0.0</stddev>
+        </noise>
+      </ray>
+      <plugin name="gazebo_ros_head_hokuyo_controller" filename="libgazebo_ros_gpu_laser.so">
```
Is ROS a required dependency for this plugin?

xabierolaz replied:
~~All lasers in gazebo seem to be related to ROS~~ Im doing further research atm if there is any alternative out of the ROS ecosystem

EDIT:

Found [GpuRaySensor](http://osrf-distributions.s3.amazonaws.com/gazebo/api/8.2.0/classgazebo_1_1sensors_1_1GpuRaySensor.html) Class Reference, researching out how to implement in the SDF model

Seems to be inside plugins folder in Gazebo 10
https://github.com/osrf/gazebo/blob/6fd426b3949c4ca73fa126cde68f5cc4a59522eb/plugins/GpuRayPlugin.cc

Now it's correctly added into the sdf model.

In case we need to use the gpu instead the cpu, just change 
` <sensor name="laser" type="ray">
`
to 
` <sensor name="laser" type="gpu_ray">
`

wil3 replied:
Doing it this way won't work, the sensor plugin you use needs to use the interface to GymFC. If the sensor plugin doesn't publish to a topic GymFC is listening to, itll never get the values. Look here, https://github.com/wil3/gymfc#api and https://github.com/wil3/gymfc-aircraft-plugins/blob/master/src/gazebo_imu_plugin.cpp#L129 for examples.

wil3 commented on gymfc/envs/assets/gazebo/plugins/msgs/State.proto:
```diff
@@ -24,15 +24,20 @@ message State
   optional float  vbat_current = 12;
 
+  // Distance Sensor
+  repeated float  distance_sensor = 13 [packed=true];
```
lets rename this to `distance`, its just the state we care about not the sensor. others should be refactored in the future.

in the comment please indicate how units are defined and how they are indexed

xabierolaz committed:
Update State.proto

deprecated distance terminology

xabierolaz committed:
Update State.proto

Deprecated sensor terminology

xabierolaz committed:
Update model.sdf

Added Gazebo GpuRaySensor, can be switched to gpu by changing 
<sensor name="laser" type="ray"> to <sensor name="laser" type="gpu_ray">

wil3 commented on gymfc/envs/assets/gazebo/plugins/msgs/State.proto:
```diff
@@ -24,15 +24,20 @@ message State
   optional float  vbat_voltage = 11;
   optional float  vbat_current = 12;
 
+  // Distance. Units will be in meters as in Gazebo, with two decimals indicating decimeters and centimeters. e.g. 4.85 m (4 meters and 85 centimeters)
```
Wrap all lines longer than 80 characters. 
This is too verbose, anyone using this plugin should know what a decimal number is. My previous comment was in regards to you using the "repeated" keyword. If this can have more than one value, does the user need to know anything about how the distance values will be ordered? Perhaps not and the comment should indicate this.

xabierolaz replied:
Understood

wil3 commented:
I suggest for this PR you scope it to just the core GymFC changes, so in addition to the state.proto this includes modifications to the cpp plugin https://github.com/wil3/gymfc/blob/master/gymfc/envs/assets/gazebo/plugins/FlightControllerPlugin.cpp. You can use IMU or ESC as an example. 

A separate PR should be used to introduce the implementation in an actual model.

xabierolaz committed:
Update State.proto

xabierolaz commented:
> I suggest for this PR you scope it to just the core GymFC changes, so in addition to the state.proto this includes modifications to the cpp plugin https://github.com/wil3/gymfc/blob/master/gymfc/envs/assets/gazebo/plugins/FlightControllerPlugin.cpp. You can use IMU or ESC as an example.
> 
> A separate PR should be used to introduce the implementation in an actual model.

Will add cpp plugin to this PR and commit sdf model in a separate PR as you stated

xabierolaz committed:
Update State.proto

xabierolaz reviewed (commented):
removed "repeated" as its single value

xabierolaz reviewed (commented):
removed repeated as its single value

xabierolaz commented:
> I suggest for this PR you scope it to just the core GymFC changes, so in addition to the state.proto this includes modifications to the cpp plugin https://github.com/wil3/gymfc/blob/master/gymfc/envs/assets/gazebo/plugins/FlightControllerPlugin.cpp. You can use IMU or ESC as an example.
> 
> A separate PR should be used to introduce the implementation in an actual model.

@wil3 distance sensor should be under the IMU category right?

wil3 commented:
> removed "repeated" as its single value

I would keep this as repeated, it gives you flexibility to return distances in the future for  different directions. It would be useful for navigation to know the distance around you, not just below you. 

What I meant by my previous comment is that a comment should be added to explain to the user the different indices could represent distances in different directions.

wil3 commented:
> > I suggest for this PR you scope it to just the core GymFC changes, so in addition to the state.proto this includes modifications to the cpp plugin https://github.com/wil3/gymfc/blob/master/gymfc/envs/assets/gazebo/plugins/FlightControllerPlugin.cpp. You can use IMU or ESC as an example.
> > A separate PR should be used to introduce the implementation in an actual model.
> 
> @wil3 distance sensor should be under the IMU category right?

Not sure what you mean by IMU category, can you be more specific?

xabierolaz committed:
Update State.proto

xabierolaz commented:
> > > I suggest for this PR you scope it to just the core GymFC changes, so in addition to the state.proto this includes modifications to the cpp plugin https://github.com/wil3/gymfc/blob/master/gymfc/envs/assets/gazebo/plugins/FlightControllerPlugin.cpp. You can use IMU or ESC as an example.
> > > A separate PR should be used to introduce the implementation in an actual model.
> > 
> > 
> > @wil3 distance sensor should be under the IMU category right?
> 
> Not sure what you mean by IMU category, can you be more specific?

@wil3  what I mean is if for having working in GymFC the distance sensor has to be in the **FlightControllerPlugin.cpp** added as an imu or as an esc value inside the code, or if its a totally different category like this 


```
#include "MotorCommand.pb.h"
#include "EscSensor.pb.h"
#include "Imu.pb.h"
#include "State.pb.h"
#include "Action.pb.h"
#include "Distance.pb.h"<<<<<
```

Another example of what I'm talking about
 
```
enum Sensors {
    IMU,
    ESC,
    BATTERY
    DISTANCE << is this way ok or distance should be considered as an Imu or Esc sensor 
  };
```


I'm creating a separate PR for the actual implementation of the model as you suggested

xabierolaz reviewed (commented):
Added as comments indexes for different directions

xabierolaz committed:
Delete model.sdf

xabierolaz committed:
Revert "Delete model.sdf"

This reverts commit 09c3fd84435bd61ff9ba2cecb2f3b2bb439cb7e7.

xabierolaz committed:
Update .gitignore
